buildSrc/src/main/java/org/elasticsearch/gradle/precommit/FilePermissionsTask.java

```diff
@@ -0,0 +1,112 @@
+import org.gradle.api.tasks.InputFiles;
+import org.gradle.api.tasks.OutputFile;
+import org.gradle.api.tasks.SourceSetContainer;
+import org.gradle.api.tasks.StopExecutionException;
+import org.gradle.api.tasks.TaskAction;
+import org.gradle.api.tasks.util.PatternFilterable;
+import org.gradle.api.tasks.util.PatternSet;
+
+/**
+ * Checks source files for correct file permissions.
+ */
+public class FilePermissionsTask extends DefaultTask {
+
+    /**
+     * A pattern set of which files should be checked.
+     */
+    private final PatternFilterable filesFilter = new PatternSet()
+            // we always include all source files, and exclude what should not be checked
+            .include("**")
+            // exclude sh files that might have the executable bit set
+            .exclude("**/*.sh");
+
+    private File outputMarker = new File(getProject().getBuildDir(), "markers/filePermissions");
+
+    public FilePermissionsTask() {
+        setDescription("Checks java source files for correct file permissions");
+    }
+
+    private static boolean isExecutableFile(File file) {
+        try {
+            Set<PosixFilePermission> permissions = Files.getFileAttributeView(file.toPath(), PosixFileAttributeView.class)
+                    .readAttributes()
+                    .permissions();
+            return permissions.contains(PosixFilePermission.OTHERS_EXECUTE)
+                    || permissions.contains(PosixFilePermission.OWNER_EXECUTE)
+                    || permissions.contains(PosixFilePermission.GROUP_EXECUTE);
+        } catch (IOException e) {
+            throw new IllegalStateException("unable to read the file " + file + " attributes", e);
+        }
+    }
+
+    /**
+     * Returns the files this task will check
+     */
+    @InputFiles
+    public FileCollection getFiles() {
+        SourceSetContainer sourceSets = getProject().getConvention().getPlugin(JavaPluginConvention.class).getSourceSets();
+        return getProject().files(sourceSets.stream()
+                .map(sourceSet -> sourceSet.getAllSource().matching(filesFilter))
+                .toArray());
+    }
+
+    @TaskAction
+    public void checkInvalidPermissions() throws IOException {
+        if (Os.isFamily(Os.FAMILY_WINDOWS)) {
+            throw new StopExecutionException();
+        }
+        List<String> failures = getFiles().getFiles().stream()
+                .filter(FilePermissionsTask::isExecutableFile)
+                .map(file -> "Source file is executable: " + file)
+                .collect(Collectors.toList());
+
+        if (!failures.isEmpty()) {
+            throw new GradleException("Found invalid file permissions:\n" + String.join("\n", failures));
+        }
+
+        Files.write(outputMarker.toPath(), "done".getBytes("UTF-8"));
+    }
+
+    @OutputFile
+    public File getOutputMarker() {
+        return outputMarker;
+    }
+
+    public void setOutputMarker(File outputMarker) {
```

Comment:
We don't need this to be configurable, it can be considered an implementation detail.  
Sorry, I missed this in my earlier review.

Reply:
I dropped it, and found out that I needed to create the missing parent folders (my previous test used an output file at the root of the TemporaryFolder). I added a .mkdirs() in the Task.